### PR TITLE
feat: add proposal for dispose functionality

### PR DESCRIPTION
**Pull request by weyert**
Add  the ability to dispose of resources consumed by the `OpenFeature Provider`. For example, when a provider uses a `setInterval` or registers an event handler then they should be released and currently this is not possible

### CODE REVIEW

File: 009-OFEP-add-dispose.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+The function can be called by the the `Client` instance when it requests the resoures need to be disposed off, for this reason, 
+the `Client`-type needs to be enhanced to also have the `dispose`-function:
+
+```typescript
+dispose(): Promise<void>
+```
+
+A potential implementation could be the following in the Node SKD:
```
Comment (justinabrahms) [inline 009-OFEP-add-dispose.md:36]:
```suggestion
A potential implementation could be the following in the Node SDK:
```

File: 009-OFEP-add-dispose.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+### [Example] Implementation
+
+The `Provider` type gets extended with the following function:
+
+```typescript
+dispose(): Promise<void>
+```
+
+The function can be called by the the `Client` instance when it requests the resoures need to be disposed off, for this reason, 
+the `Client`-type needs to be enhanced to also have the `dispose`-function:
```
[the remark this answers is not in the document]
Comment (toddbaert) [reply 009-OFEP-add-dispose.md:30]: I think I agree with @justinabrahms 's implication; it makes sense for this to happen globally, on the OpenFeatureAPI.

[the remark this answers is not in the document]
Comment (toddbaert) [reply 009-OFEP-add-dispose.md:30]: > I have to double check in the spec/code if you can make multiple Client instances or not,

Clients are lightweight, you can make more than one.

**Comment by beeme1mr** [comment]
Hey @weyert, just to confirm, the application author would need to call the `dispose` method themselves, right? I think this would work fine in most cases. However, OpenFeature should likely call `dispose` on the previously registered provider when a new one is set.

**Comment by beeme1mr** [comment]
@weyert, could you please update your proposal with the suggestions mentioned above? Also, please remember to sign-off on your comments. We're not able to merge without it.

**Comment by weyert** [comment]
Sure, hope to catch up on it next week

**Comment by beeme1mr** [comment]
@weyert, this will likely need to be something we tackle post KubeCon. However, since it should be a non-breaking change, that shouldn't be an issue.

**Comment by beeme1mr** [comment]
Hey @weyert, would you be able to update the proposal based on the feedback and sign-off on the commits?

**Comment by weyert** [comment]
> Hey @weyert, would you be able to update the proposal based on the feedback and sign-off on the commits?

Sorry, more important work stuff got in the way. I will make some time this weekend to work on it.

**Comment by toddbaert** [comment]
[This issue](https://github.com/open-feature/spec/issues/163) adds some relevant discussion to this OFEP. I will attempt to address some of the excellent points made in that issue here to keep discussion centralized. I'll try to keep my thoughts concise.

- I think both https://github.com/open-feature/spec/issues/163 and this OFEP suggest adding the shutdown method to the provider. I think instead it should be added to the top-level API object, (ex: `OpenFeatureAPI.shutdown()`). My reasoning for this is because this would shut down the provider, which would impact _all_ clients. Because the effect is global, I think it should be implemented on the global object.
- I think when `shutdown` is called, if a provider throws or otherwise abnormally functions, clients should essentially behave as they would in any case when a provider doesn't resolve values as expected, and return the default.
- We may want to implement another function to "re-initialize" the configured provider, which would again put it in a ready state.
- Implicit in the above is that the state of "readiness" (shutdown or not) is maintained in primarily in the provider, which I think is the optimal approach.

All the above seems compatible (though not completely implemented) with https://github.com/open-feature/ofep/pull/25, and with the [eventing-POC](https://github.com/open-feature/js-sdk/pull/316) in the js-sdk. I could attempt to do an implementation of the above there if it would be helpful.

cc @weyert @beeme1mr @justinabrahms @moredip @jakedoublev @skyerus

**Comment by beeme1mr** [comment]
> * We may want to implement another function to "re-initialize" the configured provider, which would again put it in a ready state.

This doesn't seem necessary. You could just reregister a provider to reinitialize a provider that was shutdown.

**Comment by toddbaert** [comment]
> This doesn't seem necessary. You could just reregister a provider to reinitialize a provider that was shutdown.

I had the same thought. We could do it this way, but re-registering the same provider feels a bit odd. Maybe it actually makes the most sense though. I don't feel too strongly about that one either way.

**Comment by jakedoublev** [comment]
> * I think both [Support in spec for closure/shutdown spec#163](https://github.com/open-feature/spec/issues/163) and this OFEP suggest adding the shutdown method to the `client`. I think instead it should be added to the top-level API object, (ex: `OpenFeatureAPI.shutdown()`). My reasoning for this is because this would shut down the provider, which would impact _all_ clients. Because the effect is global, I think it should be implemented on the global object.

This is really interesting. I hadn't experienced a use case of multiple `clients` simultaneously consuming one `provider` and one `OpenFeatureAPI` but your suggestion makes sense in light of that capability and use case. I agree that it makes the most sense to ensure a `shutdown` call closes all clients.

**Comment by tcarrio** [comment]
> > * I think both [Support in spec for closure/shutdown spec#163](https://github.com/open-feature/spec/issues/163) and this OFEP suggest adding the shutdown method to the `client`. I think instead it should be added to the top-level API object, (ex: `OpenFeatureAPI.shutdown()`). My reasoning for this is because this would shut down the provider, which would impact _all_ clients. Because the effect is global, I think it should be implemented on the global object.
> 
> This is really interesting. I hadn't experienced a use case of multiple `clients` simultaneously consuming one `provider` and one `OpenFeatureAPI` but your suggestion makes sense in light of that capability and use case. I agree that it makes the most sense to ensure a `shutdown` call closes all clients. 

OpenFeature is currently designed that you can only have a single instance of the API, and one or more clients. So the case of many clients will need to be handled.

**Comment by kinyoklion** [comment]
> > This doesn't seem necessary. You could just reregister a provider to reinitialize a provider that was shutdown.
> 
> I had the same thought. We could do it this way, but re-registering the same provider feels a bit odd. Maybe it actually makes the most sense though. I don't feel too strongly about that one either way.

I think registering a provider makes sense. I am doubtful registering the same provider (as in same instance) would make sense, as the provider would already be closed, so you would still need some means to tell it not to be closed. (The way my provider currently works, if it were to be closed somehow, then it couldn't really be started again.)

**Comment by toddbaert** [comment]
> I think registering a provider makes sense. I am doubtful registering the same provider (as in same instance) would make sense, as the provider would already be closed, so you would still need some means to tell it not to be closed.

Great point. If some providers implement a means of "re-initialization", I think that's their business, and we can leave the SDK out of it. The SDK will simply use whatever provider is set, and if the state of readiness is maintained with the provider, we don't need to define "re-initialization" behavior on the SDK.

**Comment by toddbaert** [comment]
@jakedoublev with regards to multiple clients, [this bit of doc](https://docs.openfeature.dev/docs/reference/concepts/evaluation-api#creating-a-client) my be helpful.

File: 009-OFEP-add-dispose.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
+## Add dispose functionality to Client
+
+## State: DRAFTING
+
+The goal of this OFEP is to enhance the OpenFeature Client to allow cleanly dispose of any resources by consumed by OpenFeature providers.
```
Comment (toddbaert) [inline 009-OFEP-add-dispose.md:5]:
```suggestion
## Add dispose functionality to API

## State: DRAFTING

The goal of this OFEP is to enhance the OpenFeature API to allow authors to cleanly dispose of any resources by consumed by OpenFeature providers.
```

**Comment by beeme1mr** [comment]
Hey @weyert, would you be able to address the feedback and sign off your commits? This is a pre-req for some upcoming client-side changes. Thanks!

**Comment by weyert** [comment]
> Hey @weyert, would you be able to address the feedback and sign off your commits? This is a pre-req for some upcoming client-side changes. Thanks!

I have applied the suggestions and signed them off. I will have a closer look what is needs to be done based on the comments

**Comment by toddbaert** [comment]
@weyert if you give me access to your fork, I can push the contents [here](https://github.com/open-feature/ofep/pull/46) up to it, and then we can revive this PR. Let me know!

**Pull request closed without merging**